lib/IRGen/IRGenDebugInfo.cpp

```diff
@@ -1603,6 +1668,42 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
       auto L = getFilenameAndLocation(*this, Decl);
       auto *File = getOrCreateFile(L.filename);
       unsigned FwdDeclLine = 0;
+      if (Opts.DebugInfoLevel > IRGenDebugInfoLevel::ASTTypes) {
+        // To emit full debug info for generic types, the strategy is to emit
+        // full debug info for the type with archetypes, and still emit opaque
+        // debug information for the specialized type. For example, given:
+        // struct Pair<T, U> {
+        //   let t : T
+        //   let u: U
+        // }
+        // When emitting debug information for a type such as Pair<Int, Double>,
+        // emit full debug info for Pair<T, U>, and emit the regular debug
+        // information for Pair<Int, Double>.
+
+        // Go from Pair<Int, Double> to Pair<T, U>.
+        auto UnsubstitutedTy = Decl->getDeclaredInterfaceType();
+        UnsubstitutedTy = Decl->mapTypeIntoContext(UnsubstitutedTy);
+
+        auto DbgTy = DebugTypeInfo::getFromTypeInfo(
+            UnsubstitutedTy, IGM.getTypeInfoForUnlowered(UnsubstitutedTy), IGM,
+            false);
+        Mangle::ASTMangler Mangler;
+        std::string DeclTypeMangledName = Mangler.mangleTypeForDebugger(
+            UnsubstitutedTy->mapTypeOutOfContext(), {});
+        if (DeclTypeMangledName == MangledName) {
+          return createUnsubstitutedGenericStructType(
+              DbgTy, Decl, UnsubstitutedTy, Scope, File, FwdDeclLine,
+              SizeInBits, AlignInBits, Flags, nullptr, llvm::dwarf::DW_LANG_Swift,
+              DeclTypeMangledName);
+        }
+        // Force the creation of the unsubstituted type, don't create it
+        // directly so it goes through all the caching/verification logic.
+        getOrCreateType(DbgTy);
```

Comment:
What holds on to this type so it stays alive in the IR?

Reply:
I thought `getOrCreateType` storing it in the type cache would be enough to keep it alive?

```
    DITypeCache.insert({DbgTy.getType(), llvm::TrackingMDNodeRef(DITy)});
```

Reply:
@adrian-prantl the type was only being retained because I happened to be using one of its method in the test program (for example, its init). I changed this to always forcibly retain the type instead.